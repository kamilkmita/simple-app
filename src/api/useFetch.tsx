import { useEffect, useState } from "react";
import messages from "../constants/messages";
import { useToasts } from "react-toast-notifications";

export interface Items {
  [key: string]: string | undefined;
  id: string;
  name: string;
  venue?: string;
  website?: string;
}

interface UseFetchResponseType {
  status: string;
  data: Items[];
}

export enum LoadingStatus {
  Waiting = "waiting",
  Fetching = "fetching",
  Fetched = "fetched",
}

// temporary solution for display loading feature
const sleep = (time: number) =>
  new Promise((resolve) => setTimeout(resolve, time));

const useFetch = (): UseFetchResponseType => {
  const [status, setStatus] = useState<LoadingStatus>(LoadingStatus.Waiting);
  const [data, setData] = useState<Items[]>([]);
  const { addToast } = useToasts();

  useEffect(() => {
    const fetchData = async () => {
      try {
        setStatus(LoadingStatus.Fetching);
        const response = await fetch(
          "http://api.football-data.org/v2/competitions/2021/teams/",
          {
            headers: {
              "X-Auth-Token": "7c91c189cea14e4ba605e1b171455cdf",
            },
          }
        );

        // temporary solution for display loading feature
        await sleep(2000);

        const data = await response.json();
        if (!response.ok) {
          throw response;
        }

        setData(data.teams);
        setStatus(LoadingStatus.Fetched);
      } catch {
        addToast(messages.fetchResponse, { appearance: "error" });
      }
    };
    fetchData();
  }, [addToast]);

  return { status, data };
};

export default useFetch;
