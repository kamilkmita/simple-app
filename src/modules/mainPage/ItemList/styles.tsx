import styled from "styled-components";

export const Wrapper = styled.div`
  display: flex;
  flex-wrap: wrap;
`;

export const Loader = styled.div`
  width: 100%;
  margin: 50px;
  text-align: center;
`;
