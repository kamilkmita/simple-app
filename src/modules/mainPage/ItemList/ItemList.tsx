import CircularProgress from "@material-ui/core/CircularProgress";
import useFetch, { LoadingStatus } from "../../../api/useFetch";
import Item from "../Item/Item";
import * as S from "./styles";

const ItemList = () => {
  const { status, data } = useFetch();

  const items = data.map((item) => <Item key={item.id} {...item} />);

  return (
    <S.Wrapper>
      {status === LoadingStatus.Fetching ? (
        <S.Loader>
          <CircularProgress />
        </S.Loader>
      ) : (
        items
      )}
    </S.Wrapper>
  );
};

export default ItemList;
