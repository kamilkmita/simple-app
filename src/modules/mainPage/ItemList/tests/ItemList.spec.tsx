import renderer from "react-test-renderer";
import ItemList from "../ItemList";
import useFetch, { Items, LoadingStatus } from "../../../../api/useFetch";

jest.mock("../../item/Item");
jest.mock("../../../../api/useFetch");

const mockData: Items[] = [
  {
    id: "1",
    name: "item1Name",
    venue: "item1Venue",
    website: "item1Website",
  },
  {
    id: "2",
    name: "item2Name",
    venue: "item2Venue",
  },
  {
    id: "3",
    name: "item3Name",
    website: "item1Website",
  },
];

describe("ItemList", () => {
  test("when data is not fetched yet", () => {
    (useFetch as jest.Mock).mockReturnValue({
      status: LoadingStatus.Fetching,
      data: [],
    });

    const component = renderer.create(<ItemList />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test("when data is empty", () => {
    (useFetch as jest.Mock).mockReturnValue({
      status: LoadingStatus.Fetched,
      data: [],
    });

    const component = renderer.create(<ItemList />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  test("when data is filled", () => {
    (useFetch as jest.Mock).mockReturnValue({
      status: LoadingStatus.Fetched,
      data: mockData,
    });

    const component = renderer.create(<ItemList />);

    expect(component.toJSON()).toMatchSnapshot();
  });
});
