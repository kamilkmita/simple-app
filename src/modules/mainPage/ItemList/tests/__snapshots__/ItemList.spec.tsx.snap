// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ItemList when data is empty 1`] = `
<div
  className="sc-bdvvaa MoxAp"
/>
`;

exports[`ItemList when data is filled 1`] = `
<div
  className="sc-bdvvaa MoxAp"
>
  ItemMockedComponent
  ItemMockedComponent
  ItemMockedComponent
</div>
`;

exports[`ItemList when data is not fetched yet 1`] = `
<div
  className="sc-bdvvaa MoxAp"
>
  <div
    className="sc-gsDJrp fyNeOs"
  >
    <div
      className="MuiCircularProgress-root MuiCircularProgress-colorPrimary MuiCircularProgress-indeterminate"
      role="progressbar"
      style={
        Object {
          "height": 40,
          "width": 40,
        }
      }
    >
      <svg
        className="MuiCircularProgress-svg"
        viewBox="22 22 44 44"
      >
        <circle
          className="MuiCircularProgress-circle MuiCircularProgress-circleIndeterminate"
          cx={44}
          cy={44}
          fill="none"
          r={20.2}
          strokeWidth={3.6}
          style={Object {}}
        />
      </svg>
    </div>
  </div>
</div>
`;
